frontend/discourse/app/lib/user-presence.js

```diff
@@ -98,9 +105,18 @@ function processChanges() {
 }
 
 export function seenUser() {
-  userSeenJustNow = true;
-  if (callbackWaitingForPresence) {
-    processChanges();
+  // a boolean check is going to be 10x to 80x faster than Date.now()
+  // scroll, touchmove, click, keydown can all happen very frequently
+  // this debounces to de-risk
+  if (!waitingOnAnimationFrame) {
+    waitingOnAnimationFrame = true;
+    requestAnimationFrame(() => {
+      waitingOnAnimationFrame = false;
+      lastUserActivity = Date.now();
+      if (callbackWaitingForPresence) {
+        processChanges();
+      }
+    });
```

Comment:
The `callbackWaitingForPresence` check means that the cost of this was already just a boolean comparison in 99.99% of cases. (`callbackWaitingForPresence` is set false in `processChanges()`)

`requestAnimationFrame` is intended for DOM updates, so it's a little strange to schedule 'data layer' stuff there. It means that painting frames to the screen will be delayed while presence callbacks are run.

Was this change to solve a particular perf problem? If so do you have a repro you can share?

Reply:
Previously though there was an extra arbitrary delay to discover a user was not present 10-20 seconds, this allows us to get the info right away  with minimal delays. that said, I can be explicit here on the debounce and give it 1 second fidelity, of accept chunks of 10 calls to Date.now, scroll and touchmove are the biggest concerns here.

What is your pref here:

- accept new Date.now calls
- debounce with a setTimeout instead and go for 1 second fidelity
- keep the old, we only boolean in the noisy event and live with the 10-20 sec delay on discovering a user is not present.